### PR TITLE
More integration tests

### PR DESCRIPTION
**Pull request by tyll**
Add more integration tests and scripts to gather coverage data from integration tests

### CODE REVIEW

**Comment by tyll** [comment]
The Fedora 26 CI check failure is unexpected. The log at

https://larsu.fedorapeople.org/logs/linux-system-roles-network-pull-50-3af11c8-fedora-26-20180518-213512/test.log

ends with:

`TASK [linux-system-roles.network : Configure networking connection profiles] ***
task path: /tmp/tmp35497jy9/tests/roles/linux-system-roles.network/tasks/main.yml:21
fatal: [/cache/Fedora-Cloud-Base-26-1.5.x86_64.qcow2]: FAILED! => {"changed": false, "failed": true, "msg": "fatal error: failure loading libnm library: No module named gi"}
	to retry, use: --limit @/tmp/tmp35497jy9/tests/tests_ethernet.retry`

Not sure what is wrong here, any hints @thom311 ?

**Comment by tyll** [comment]
This includes now commits from PR #54 to fix CI results, I will rebase once #54 is merged.

**Comment by thom311** [comment]
the role uses Python + libnm + pygobject. Hence it needs some python+gobject-integration library...

Not sure which. It always confuses me. I think the package name also depends on the distibution.

How about `python3-gobject-base` package?

**Comment by tyll** [comment]
@thom311 do you know how it works on RHEL7/CentOS7, IIRC there is no python-gobject-base installed by default. Should ```python3-gobject-base``` be added to the packages in the tasks file for newer systems?

**Comment by thom311** [comment]
First of all, the pygobject dependency is only necessary if you use the nm provider. initscripts don't require it (e.g. RHEL6).

I guess yes, the right (which??) package should be included to the required packages list, with nm-provider.

**Comment by tyll** [comment]
It seems that the required package is `python{,3}-gobject-base`.

File: tests/get-coverage.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        coverage: /root/.local/bin/coverage
+    - name: set more facts
+      set_fact:
+        coverage_file: ansible-coverage-{{ coverage_module }}-{{ test_playbook|replace('.yml', '') }}
+
+    - name: set more facts
+      set_fact:
+        destdir: "remote_coverage/{{ test_playbook }}"
+      when: destdir is undefined
+
```
Comment (pcahyna) [inline tests/get-coverage.yml]: ditto. A variable set in `vars:` will be overriden by `-e `, so `when:` should not be needed.
http://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

File: tests/get-coverage.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  hosts: all
+  tasks:
+    - name: set facts
+      set_fact:
+        coverage_module: network_connections
+        coverage: /root/.local/bin/coverage
+    - name: set more facts
+      set_fact:
+        coverage_file: ansible-coverage-{{ coverage_module }}-{{ test_playbook|replace('.yml', '') }}
+
```
Comment (pcahyna) [inline tests/get-coverage.yml]: Can't you just use `vars:` at the top of the play?

File: tests/get-coverage.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    - name: remove old data
+      file:
+        state: absent
+        path: "{{ item }}"
+      loop:
+        - .coverage
+
+    - name: remove test data
+      shell: rm -rf /root/.ansible/ /root/.coverage.* /root/htmlcov /tmp/ansible_*
+
```
Comment (pcahyna) [inline tests/get-coverage.yml]: can this be merged with `file` above?

File: tests/get-coverage.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      pip:
+        name: coverage
+        extra_args: --user --upgrade
+
+    - name: remove old data
+      file:
+        state: absent
+        path: "{{ item }}"
+      loop:
+        - .coverage
```
Comment (pcahyna) [inline tests/get-coverage.yml]: again? why?

File: tests/get-coverage.yml (hunk 3 above)
Comment (tyll) [reply tests/get-coverage.yml]: The problem is the wildcards, AFAICS `with_fileglob` only works locally. I guess we only need `.coverage.*` anymore, the other paths are leftovers from my testing.

File: tests/get-coverage.yml (hunk 2 above)
Comment (tyll) [reply tests/get-coverage.yml]: The problem with using `vars:` is that the variables are only set for the current playbook but I need it in several playbooks:

```
---
- name: set var
  hosts: all
  vars:
    - example_var: foo

- name: Show the var
  hosts: all
  tasks:
    - debug:
        var: example_var
```

```
ansible-playbook -i rhel7, test_things.yml

PLAY [set var] ****************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [rhel7]

PLAY [Show the var] ***********************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [rhel7]

TASK [debug] ******************************************************************************************************************************************************************************************************
ok: [rhel7] => {
    "example_var": "VARIABLE IS NOT DEFINED!"
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
rhel7                      : ok=3    changed=0    unreachable=0    failed=0
```

File: tests/get-coverage.yml (hunk 1 above)
Comment (tyll) [reply tests/get-coverage.yml]: I removed it.

File: tests/get-coverage.yml (hunk 4 above)
Comment (tyll) [reply tests/get-coverage.yml]: The file is removed to ensure that we get fresh coverage data.

File: tests/get-coverage.yml (hunk 2 above)
Comment (pcahyna) [reply tests/get-coverage.yml]: can you please add a comment explaining whi it needs to be this way? Also, why have two `set_facts:` ?

Comment (tyll) [reply tests/get-coverage.yml]: Done

File: tests/get-coverage.yml (hunk 4 above)
Comment (pcahyna) [reply tests/get-coverage.yml]: yes, but why twice? (see line 31). Does any operation in-between recreate it?

File: tests/get-coverage.yml [hunk 5]
```diff
@@ -1,19 +1,22 @@
 # SPDX-License-Identifier: BSD-3-Clause
 ---
+# This expects the variable test_playbook to be set from the outside
 - name: Prepare for coverage extraction
   hosts: all
   tasks:
+    # Use set_fact to set variables to make them available in all playbooks
+    # vars: Would only set variables for the current playbook
```
Comment (pcahyna) [inline tests/get-coverage.yml]: "all plays" and "current play" is probably more accurate, as your example shows.

Comment (tyll) [reply tests/get-coverage.yml]: fixed

File: tests/get-coverage.yml (hunk 4 above)
Comment (tyll) [reply tests/get-coverage.yml]: Good catch, it was wrong. Also it should have used the variable. It is all fixed and squashed now.

**Comment by tyll** [comment]
Thank you for the review.

**Pull request merged**